python_modules/dagster/dagster/core/definitions/sensor_definition.py

```diff
@@ -602,3 +602,236 @@ def _fn(context):
+            run should be launched or not. Takes a :py:class:`~dagster.SensorEvaluationContext` and
+            a list of :py:class:`~dagster.core.storage.event_log.EventLogEntry` corresponding to
+            :py:class:`~dagster.AssetMaterialization` events for all of the asset keys monitored
+            by this sensor.
+
+            This function must return a generator, which must yield either a single
+            :py:class:`~dagster.SkipReason` or one or more :py:class:`~dagster.RunRequest` objects.
+        job (Optional[Union[GraphDefinition, JobDefinition]]): The job object to target with this
+            sensor.
+        jobs (Optional[Sequence[Union[GraphDefinition, JobDefinition]]]): (experimental) A list of
+            jobs to be executed when the sensor fires.
+        minimum_interval_seconds (Optional[int]): The minimum number of seconds that will elapse
+            between sensor evaluations.
+        description (Optional[str]): A human-readable description of the sensor.
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        asset_keys: List[AssetKey],
+        asset_materialization_fn: Callable[
+            ["SensorExecutionContext", List["EventLogEntry"]],
+            Union[Generator[Union[RunRequest, SkipReason], None, None], RunRequest, SkipReason],
+        ],
+        job: Optional[Union[GraphDefinition, JobDefinition]] = None,
+        jobs: Optional[Sequence[Union[GraphDefinition, JobDefinition]]] = None,
+        minimum_interval_seconds: Optional[int] = None,
+        description: Optional[str] = None,
+    ):
+        check.invariant(
+            asset_keys != [], "Must not provide empty asset_keys when defining a multi asset sensor"
+        )
+        self._asset_keys = check.list_param(asset_keys, "asset_keys", of_type=AssetKey)
+
+        from dagster.core.events import DagsterEventType
+        from dagster.core.events.log import EventLogEntry
+        from dagster.core.storage.event_log.base import EventRecordsFilter
+
+        def _wrap_asset_fn(
+            materialization_fn: Callable[
+                [SensorEvaluationContext, List[EventLogEntry]],
+                Union[Generator[Union[RunRequest, SkipReason], None, None], RunRequest, SkipReason],
+            ]
+        ):
+            def _fn(context: SensorEvaluationContext):
+                after_cursor = [None for _ in self.asset_keys]
+
+                # This may be overly defensive
+                if context.cursor:
+                    try:
+                        after_cursor = json.loads(context.cursor)
+                    except ValueError:
+                        pass
+
+                # TODO: This won't be performant for large numbers of assets (we'll be doing N
+                # roundtrips to the database, so will want to extend the EventRecordsFilter API or
+                # add an additional method on the instance)
+                event_records = [
```

Comment:
what I'm not sure about is why we should go this route instead of providing a utility function for fetching the asset materialization records for multiple asset keys.  Users could then implement their own custom any/multi logic.  Is it for the cursor management?

Or is the true value that we can then query for the set of asset keys off of the sensor definition?   Illustrating how that data would be used would provide a lot of context.  I don't have a problem with this implementation (everything looks good), but I want to have a good handle on what inferences/connections we're making on the Dagit side of things.

The other question I have is would we want to attribute the sensor firing to a particular asset materialization?  Should it be the last one?  Should it be all of them?  How would we use that information?  Do we need to serialize this type of information into the tick history?

Reply:
One example where we've made the other decision is to not implement an `@s3_sensor`.  Instead we provide the `get_s3_keys` utility function, because there's no way we can cover all the ways that users want to customize their trigger logic.  Also, we're not making use of the fact that we know anything about s3 in Dagit, although I guess we could (e.g. maybe even deep linking, etc)

Reply:
These are perspicacious questions. Regardless of our ultimate aims, I think this is worth implementing and making available to users as a utility (though there might be an argument for not putting it at the top level in `dagster`) -- so I think I maybe fall on the side of "`s3_asset_sensor` is worth implementing". However, I am opening this PR in service of asset lineage. Ultimately I want it to be easy (as in the demo) to write off-the-shelf sensors with multiple upstream assets and have the lineage be correctly attributed. I'm moving slowly because of other commitments but in ordinary circumstances there would be an associated diff that adds asset sensor metadata to the asset GraphQL API and renders it in Dagit. See: https://github.com/dagster-io/dagster/commit/fb88ee9b3d2691720e499de80ff7b70f5057c227, https://github.com/dagster-io/dagster/commit/3b8b6e91c042e54f71a9fdc0c5e973677af19bca, https://github.com/dagster-io/dagster/commit/c369641dd5f10f68f44484fd85a12965d95b55db.

Reply:
Yeah, absolutely.  I've just erred on the side of being very conservative w.r.t. the sensor surface area until we know what types of features we will want, wish we had, etc.

One thing I implemented (and then abandoned) in a branch was to be able to attribute a run request generated from an asset sensor to the pipeline run that materialized the asset.  We could then do some deep linking in the run log (by doing some sort of engine event insertion) where the originating pipeline run has a log entry saying something like, this asset materialization kicked off a downstream run of a different pipeline.  I thought it was very cool to be able to trace through history, but it was only possible by being able to map a run request to a single asset materialization.

I'm not sure if this is what you mean by lineage (this could be a whole discussion).  But yeah, I'm all for moving forward and building useful things (even though they might not handle every case we might hypothetically want to support).

Practically speaking, do you think we should support both asset_sensor and multi_asset_sensor, or should we just support multiple assets using the existing decorator.

Reply:
I think we should have a real lineage conversation -- the thing you're describing seems incredibly useful, though I wonder if we could do it using tags (the way we do parent/child runs for reexecution) rather than by mutating the originating run's event log -- this might make it easier to have multiple parentage.

On the question of whether to just extend the existing decorator, I wasn't certain -- I think that would force the question of GraphQL support, frontend, etc. because we would probably need to change the type on `assetKey` everywhere (not sure whether there are backcompat issues involved). This is deferred by the current diff because of the instance check you can see in https://github.com/dagster-io/dagster/commit/c369641dd5f10f68f44484fd85a12965d95b55db.

Also happy to mark these as experimental.

Reply:
For the sake of user convenience, I'm supportive of offering an out-of-the-box implementation of a multi-asset sensor that covers a common pattern, even if we know that there are multi-asset sensing patterns that it does not cover.

I think it's important to make sure the name is scoped to the pattern that we're supporting.  E.g. I'd much prefer a name like `all_assets_updated_sensor` to `multi_asset_sensor`, because it gives us room in the future to add other kinds of multi-asset sensors that support different or wider patterns.

I think that the question of run-level lineage is tricky, and my hunch is that we can offer comparable value with much less headache by sticking to higher-level associations.  I.e. I think there's a decent amount of value in being able to tell users "the sensor for job X cares about materializations of asset Y", and that that would be relatively easy to implement.  I think that trying to tell users "materialization Y caused run X to be launched" is an order of magnitude trickier, especially in the multi-asset case, and not that much more valuable.

Reply:
What about `all_assets_sensor` and `any_asset_sensor`? (On analogy with Python's `all` and `any`.)

Reply:
@mgasner those names sound good to me

Reply:
@prha happy to do that refactor now or in a follow-on -- it will force the frontend issue as the graphql layer is doing an instanceof check against AssetSensorDefinition

Reply:
Yep, I think that's fine.  I think at the time we need to differentiate on the frontend, we can change the underlying definition.

Reply:
We can do the merging / changing of AssetSensorDefinition as a follow-on if you prefer, but would not want to export MultiAssetSensorDefinition from the dagster package.